smartsim/launcher/lsf/lsfLauncher.py

```diff
@@ -0,0 +1,189 @@
+            if out:
+                step_id = parse_bsub(out)
+                logger.debug(f"Gleaned batch job id: {step_id} for {step.name}")
+        else:
+            # jsrun doesn't direct output for us.
+            out, err = step.get_output_files()
+            output = open(out, "w+")
+            error = open(err, "w+")
+            task_id = self.task_manager.start_task(
+                cmd_list, step.cwd, out=output, err=error
+            )
+
+        # if batch submission did not successfully retrieve job ID
+        if not step_id and step.managed:
+            step_id = self._get_lsf_step_id(step)
+        self.step_mapping.add(step.name, step_id, task_id, step.managed)
+        return step_id
+
+    def stop(self, step_name):
+        """Stop/cancel a job step
+
+        :param step_name: name of the job to stop
+        :type step_name: str
+        :return: update for job due to cancel
+        :rtype: StepInfo
+        """
+        stepmap = self.step_mapping[step_name]
+        if stepmap.managed:
+            qdel_rc, _, err = bkill([str(stepmap.step_id)])
+            if qdel_rc != 0:
+                logger.warning(f"Unable to cancel job step {step_name}\n {err}")
+            if stepmap.task_id:
+                self.task_manager.remove_task(stepmap.task_id)
+        else:
+            self.task_manager.remove_task(stepmap.task_id)
+
+        _, step_info = self.get_step_update([step_name])[0]
+        step_info.status = STATUS_CANCELLED  # set status to cancelled instead of failed
+        return step_info
+
+    def _get_lsf_step_id(self, step, interval=2, trials=5):
+        """Get the step_id of a step from bjobs (rarely used)
+
+        Parses bjobs output by looking for the step name
+        """
+        time.sleep(interval)
+        step_id = "unassigned"
+        username = os.environ["USER"]
+        while trials > 0:
+            output, _ = bjobs(["-w", "-u", username])
+            step_id = parse_step_id_from_bjobs(output, step.name)
+            if step_id:
+                break
+            else:
+                time.sleep(interval)
+                trials -= 1
+        if not step_id:
+            raise LauncherError("Could not find id of launched job step")
+        return step_id
+
+    def _get_managed_step_update(self, step_ids):
+        """Get step updates for WLM managed jobs
+
+        :param step_ids: list of job step ids
+        :type step_ids: list[str]
+        :return: list of updates for managed jobs
+        :rtype: list[StepInfo]
+        """
+        updates = []
+        # Include recently finished jobs
+        bjobs_args = ["-a"] + step_ids
+        bjobs_out, _ = bjobs(bjobs_args)
+        stats = [parse_bjobs_jobid(bjobs_out, str(step_id)) for step_id in step_ids]
+        # create LSFStepInfo objects to return
+
+        for stat, _ in zip(stats, step_ids):
+            info = LSFStepInfo(stat, None)
```

Comment:
Not sure what you mean here, let's discuss this in a call.

Reply:
Ok. Essentially the exitcode of the process.

e.g. in slurm we get it from ``sacct``

```text
       JobID    JobName  Partition    Account  AllocCPUS      State ExitCode 
------------ ---------- ---------- ---------- ---------- ---------- -------- 
1462295      interacti+      workq                   336    TIMEOUT             0:0 
1462295.ext+     extern                              336  COMPLETED             0:0 
```
Is there anywhere in LSF we can parse this from?

Reply:
Ok, yet another command, `bhist`. I think I can get the exit code, yes.

Reply:
Mh, no, `bhist` takes 15 seconds to run, it's now what we want. So, if we want the exit code of the single step (as the steps you launch within the same allocation, those which in your example may be 1462295.0, 1462295.1, etc), we need to keep an internal map of their task number as said in another comment. I'll give it a try, but it might be something to leave for the C API anyhow.

Reply:
Relying solely on status for now is fine. Like you said, the plan is the C API's anyway.